tsconfig.json

```diff
@@ -7,8 +7,8 @@
       "esnext"
     ],
     "moduleDetection": "force",
-    "module": "esnext",
-    "moduleResolution": "node",
+    "module": "NodeNext",
```

Comment:
Feel free to change this to Node16 if you like for 'stability', but I think NodeNext is perfectly fine.

Reply:
I would always prefer "Bundler" - I wonder if there is any downside to it?

Reply:
Yes, Bundler wouldn't have caught my runtime error, because it wouldn't have prompted you to add the .js to the end of the file.

There might be something you can tweak in your unbuild settings to make this work, but NodeNext is always going to be safer and more predictable than Bundler for libraries.

Reply:
Ah, I see the problem now. But I am honestly not a fan of the `.js` extension, especially when all files are actually `.ts`. Is it possible to have a middle ground where the library extension error can be caught but not force every import to have `.js`?

Reply:
This is a discussion of technical correctness vs aesthetics. In this case technical correctness is always the right answer.

I understand why this feels weird. I thought so too at first, but you get used to it fast, and TypeScript checks it for you.

Reply:
I wouldn't mind having the explicit extension for correctness. But things such as `import type { Options } from './options.js` bothers me because

1. `./options.js` doesn't actually exist on the disk. There is another hidden convention involved here, which, is a bit counter-intuitive to me, and doesn't contribute the explicitness much.
3. Importing types from a JavaScript file doesn't make sense with the mindset of writing JS IMO.

I am more tempted to have `allowImportingTsExtensions` to move on. It would be even better if TS could have `exports` field for `typescript/lib/tsserverlibrary` tho.

Reply:
Also for types-only imports it’s technically correct. In this case not because of Node.js, but because TypeScript dictates this.

[In TypeScript 5.3 `typescript/lib/tsserverlibrary.js` was made an alias of `typescript`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-3.html#consolidation-between-tsserverlibraryjs-and-typescriptjs). So all `'typescript/lib/tsserverlibrary'` imports can be replaced with `'typescript'`.

Reply:
@antfu Could we merge this, then look for a better solution afterwards? I am currently blocked by this, so having a patch would be pretty useful.

Reply:
@mattpocock Sorry, I am on a conf and wasn't very responsive. I am still not sure `.js` is a long-term solution for me (unless @orta has some preference). It's nice as @remcohaszing pointed out - could we have a smaller PR that changes `typescript/lib/tsserverlibrary` to `typescript` to fix the runtime error, without adding `.js` for now?

Reply:
Sure thing!

https://github.com/twoslashes/twoslash/pull/43

Reply:
Instead of adding extensions manually to all the files, a plugin like https://github.com/favware/esbuild-plugin-file-path-extensions can be created for unbuild to add the extensions at build time the you can use `Bundler` module resolution